File: src/gui4/windows/kDrive client/kDrive client/ViewModels/App/Drive.cs
/*
 * Infomaniak kDrive - Desktop
 * Copyright (C) 2023-2025 Infomaniak Network SA
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using DynamicData;
using Infomaniak.kDrive.ServerCommunication.Interfaces;
using Infomaniak.kDrive.Types;
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Collections.ObjectModel;
using System.Drawing;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Infomaniak.kDrive.ViewModels
{
    public class Drive : UISafeObservableObject, IDrive
    {
        private DbId _dbId = -1;
        private DriveId _driveId = -1;
        private string _name = "";
        private Color _color = Color.Blue;
        private bool _isFreeOffer = true; // Indicates if the drive is a free offer
        private ObservableCollection<Sync> _syncs = new ObservableCollection<Sync>();
        private Sync? _mainSync;
        private bool _isConfigured = false; // Indicates if at least one sync (which is not an advanced sync) is set up for this drive
        private bool _isAdmin = false; // Indicates if the user is admin of this drive
        private ObservableCollection<Sync> _advancedSyncs = new ObservableCollection<Sync>();

        private Account _account;
        public Drive(DbId dbId, Account account)
        {
            _dbId = dbId;
            _account = account;
            _syncs.CollectionChanged += (s, e) => RefreshAdvancedSyncsMap();

        }
        public DbId UserDbId { get => _account.User.DbId; }

        private void RefreshAdvancedSyncsMap()
        {
            var advancedSyncs = Syncs.Where(s => s != MainSync);
            foreach (int i in Enumerable.Range(0, _advancedSyncs.Count).Reverse())
            {
                Sync? sync = _advancedSyncs.ElementAt(i);
                if (!advancedSyncs.Contains(sync))
                    _advancedSyncs.Remove(sync);
            }

            foreach (var sync in advancedSyncs)
                if (!_advancedSyncs.Contains(sync))
                    _advancedSyncs.Add(sync);

            MainSync = Syncs.FirstOrDefault(s => s.RemotePath == "/" || s.RemotePath == "");
        }

        public DbId DbId
        {
            get => _dbId;
            set => SetPropertyInUIThread(ref _dbId, value);
        }

        public DriveId DriveId
        {
            get => _driveId;
            set => SetPropertyInUIThread(ref _driveId, value);
        }

        public AccountId AccountId
        {
            get => _account.AccountId;
        }

        public string AccountName
        {
            get => _account.Name;
        }

        public string Name
        {
            get => _name;
            set => SetPropertyInUIThread(ref _name, value);
        }

        public Color Color
        {
            get => _color;
            set => SetPropertyInUIThread(ref _color, value);
        }
        public bool IsAdmin
        {
            get => _isAdmin;
            set => SetPropertyInUIThread(ref _isAdmin, value);
        }

        public bool IsFreeOffer
        {
            get => _isFreeOffer;
            set => SetPropertyInUIThread(ref _isFreeOffer, value);
        }

        public ObservableCollection<Sync> Syncs
        {
            get { return _syncs; }
        }

        public Sync? MainSync
        {
            get => _mainSync;
            set
            {
                SetPropertyInUIThread(ref _mainSync, value);
                IsConfigured = value is not null;
            }
        }

        public bool IsConfigured
        {
            get => _isConfigured;
            set => SetPropertyInUIThread(ref _isConfigured, value);
        }

        public ObservableCollection<Sync> AdvancedSyncs
        {
            get => _advancedSyncs;
        }

        public Uri GetWebUri()
        {
            return App.Constants.Drive.kSuiteHomeUrl(DriveId);
        }

        public Uri GetWebTrashUri()
        {
            return App.Constants.Drive.TrashUrl(DriveId);
        }

        public Uri GetWebFavoritesUri()
        {
            return App.Constants.Drive.FavoriteUrl(DriveId);
        }

        public Uri GetWebSharedUri()
        {
            return App.Constants.Drive.SharedUrl(DriveId);
        }

        public Account Account
        {
            get => _account;
            set => SetPropertyInUIThread(ref _account, value);
        }

        public async Task RemoveSync(Sync sync, CancellationToken cancellationToken)
        {
            var commService = App.ServiceProvider.GetRequiredService<IServerCommService>();
            await commService.RemoveSync(sync.DbId, cancellationToken);
        }

    }
}

File: src/libcommon/data/drive.cpp
/*
 * Infomaniak kDrive - Desktop
 * Copyright (C) 2023-2026 Infomaniak Network SA
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "data/drive.h"

#include "libcommonserver/log/log.h"
#include "utility/utility.h"

#include <log4cplus/loggingmacros.h>

static const auto driveInfoDbId = "dbId";
static const auto driveInfoId = "id";
static const auto driveInfoAccountDbId = "accountDbId";
static const auto driveInfoName = "name";
static const auto driveInfoSize = "size";
static const auto driveInfoColor = "color";
static const auto driveInfoNotifications = "notifications";
static const auto driveInfoAdmin = "admin";
static const auto driveInfoMaintenance = "maintenance";
static const auto driveInfoLocked = "locked";
static const auto driveInfoUsedSize = "usedSize";
static const auto driveInfoAccessDenied = "accessDenied";
static const auto driveInfoPackInfo = "packInfo";

namespace KDC {

Drive::Drive(const DriveDbId dbId, const DriveId driveId, const AccountDbId accountDbId, const std::string &name,
             const int64_t size, const std::string &color, const bool notifications, const bool admin) :
    _dbId(dbId),
    _driveId(driveId),
    _accountDbId(accountDbId),
    _name(name),
    _size(size),
    _color(color),
    _notifications(notifications),
    _admin(admin) {}

void Drive::toDynamicStruct(Poco::DynamicStruct &dstruct) const {
    CommonUtility::writeValueToStruct(dstruct, driveInfoDbId, dbId());
    CommonUtility::writeValueToStruct(dstruct, driveInfoId, driveId());
    CommonUtility::writeValueToStruct(dstruct, driveInfoAccountDbId, accountDbId());
    CommonUtility::writeValueToStruct(dstruct, driveInfoName, CommonUtility::str2CommString(name()));
    CommonUtility::writeValueToStruct(dstruct, driveInfoSize, size());
    CommonUtility::writeValueToStruct(dstruct, driveInfoColor, CommonUtility::str2CommString(color()));
    CommonUtility::writeValueToStruct(dstruct, driveInfoNotifications, notifications());
    CommonUtility::writeValueToStruct(dstruct, driveInfoAdmin, admin());
    CommonUtility::writeValueToStruct(dstruct, driveInfoMaintenance,
                                      maintenanceInfo().inMaintenance()); // TODO : send all maintenance info?
    CommonUtility::writeValueToStruct(dstruct, driveInfoLocked, locked());
    CommonUtility::writeValueToStruct(dstruct, driveInfoUsedSize, usedSize());
    CommonUtility::writeValueToStruct(dstruct, driveInfoAccessDenied, accessDenied());
    CommonUtility::writeValueToStruct(dstruct, driveInfoPackInfo, packInfo(), info2DynamicVar<PackInfo>);
}

void Drive::fromDynamicStruct(const Poco::DynamicStruct &dstruct) {
    DriveDbId driveDbId = 0;
    CommonUtility::readValueFromStruct(dstruct, driveInfoDbId, driveDbId);
    setDbId(driveDbId);

    DriveId driveId = 0;
    CommonUtility::readValueFromStruct(dstruct, driveInfoId, driveId);
    setDriveId(driveId);

    AccountDbId accountDbId = 0;
    CommonUtility::readValueFromStruct(dstruct, driveInfoAccountDbId, accountDbId);
    setAccountDbId(accountDbId);

    CommString name;
    CommonUtility::readValueFromStruct(dstruct, driveInfoName, name);
    setName(CommonUtility::commString2Str(name));

    int64_t size = 0;
    CommonUtility::readValueFromStruct(dstruct, driveInfoSize, size);
    setSize(size);

    CommString color;
    CommonUtility::readValueFromStruct(dstruct, driveInfoColor, color);
    setColor(CommonUtility::commString2Str(color));

    bool notifications = false;
    CommonUtility::readValueFromStruct(dstruct, driveInfoNotifications, notifications);
    setNotifications(notifications);

    bool admin = false;
    CommonUtility::readValueFromStruct(dstruct, driveInfoAdmin, admin);
    setAdmin(admin);

    bool inMaintenance = false;
    CommonUtility::readValueFromStruct(dstruct, driveInfoMaintenance, inMaintenance);
    auto newMaintenanceInfo = maintenanceInfo();
    newMaintenanceInfo.setInMaintenance(inMaintenance);
    setMaintenanceInfo(newMaintenanceInfo);

    CommonUtility::readValueFromStruct(dstruct, driveInfoLocked, _locked);
    CommonUtility::readValueFromStruct(dstruct, driveInfoUsedSize, _usedSize);
    CommonUtility::readValueFromStruct(dstruct, driveInfoAccessDenied, _accessDenied);
    CommonUtility::readValueFromStruct(dstruct, driveInfoPackInfo, _packInfo, dynamicVar2Struct<PackInfo>);
}

void operator>>(QDataStream &in, Drive &drive) {
    qint64 dbId{0};
    qint64 id{0};
    qint64 accountDbId{0};
    QString name;
    QColor color;
    bool notifications{false};
    bool admin{false};
    bool maintenance{false};
    bool locked{false};
    bool accessDenied{false};

    in >> dbId >> id >> accountDbId >> name >> color >> notifications >> admin >> maintenance >> locked >> accessDenied;

    drive.setDbId(static_cast<DriveDbId>(dbId));
    drive.setDriveId(static_cast<DriveId>(id));
    drive.setAccountDbId(static_cast<AccountDbId>(accountDbId));
    drive.setName(name.toStdString());
    drive.setColor(color.name().toStdString());
    drive.setNotifications(notifications);
    drive.setAdmin(admin);
    MaintenanceInfo maintenanceInfo;
    maintenanceInfo.setInMaintenance(maintenance);
    drive.setMaintenanceInfo(maintenanceInfo);
    drive.setLocked(locked);
    drive.setAccessDenied(accessDenied);
}

QDataStream &operator<<(QDataStream &out, const Drive &drive) {
    out << static_cast<qint64>(drive.dbId()) << static_cast<qint64>(drive.driveId()) << static_cast<qint64>(drive.accountDbId())
        << QString::fromStdString(drive.name()) << QColor(QString::fromStdString(drive.color())) << drive.notifications()
        << drive.admin() << drive.maintenanceInfo().inMaintenance() << drive.locked() << drive.accessDenied();
    return out;
}

void operator>>(QDataStream &in, QList<Drive> &list) {
    uint64_t count = 0;
    in >> count;
    for (uint64_t i = 0; i < count; i++) {
        Drive info;
        in >> info;
        list.push_back(info);
    }
}

QDataStream &operator<<(QDataStream &out, const QList<Drive> &list) {
    auto count = static_cast<uint64_t>(list.size());
    out << count;
    for (uint64_t i = 0; i < count; i++) {
        out << list[static_cast<qsizetype>(i)];
    }
    return out;
}

} // namespace KDC
